src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -74,118 +78,58 @@
  * the rows in a replica response that don't have a corresponding row in other replica responses, and requests them by
  * primary key to the "silent" replicas in a second fetch round.
  * <p>
- * See CASSANDRA-8272 and CASSANDRA-8273 for further details.
+ * See CASSANDRA-8272, CASSANDRA-8273, and CASSANDRA-15907 for further details.
  */
 class ReplicaFilteringProtection
 {
     private static final Logger logger = LoggerFactory.getLogger(ReplicaFilteringProtection.class);
+    private static final NoSpamLogger oneMinuteLogger = NoSpamLogger.getLogger(logger, 1, TimeUnit.MINUTES);
+
+    private static final Function<UnfilteredRowIterator, EncodingStats> NULL_TO_NO_STATS =
+        rowIterator -> rowIterator == null ? EncodingStats.NO_STATS : rowIterator.stats();
 
     private final Keyspace keyspace;
     private final ReadCommand command;
     private final ConsistencyLevel consistency;
     private final InetAddress[] sources;
     private final TableMetrics tableMetrics;
 
-    /**
-     * Per-source primary keys of the rows that might be outdated so they need to be fetched.
-     * For outdated static rows we use an empty builder to signal it has to be queried.
-     */
-    private final List<SortedMap<DecoratedKey, BTreeSet.Builder<Clustering>>> rowsToFetch;
+    private final int cachedRowsWarnThreshold;
+    private final int cachedRowsFailThreshold;
+    
+    /** Tracks whether or not we've already hit the warning threshold while evaluating a partition. */
+    private boolean hitWarningThreshold = false;
+
+    private int currentRowsCached = 0; // tracks the current number of cached rows
+    private int maxRowsCached = 0; // tracks the high watermark for the number of cached rows
 
     /**
-     * Per-source list of all the partitions seen by the merge listener, to be merged with the extra fetched rows.
+     * Per-source list of the pending partitions seen by the merge listener, to be merged with the extra fetched rows.
      */
-    private final List<List<PartitionBuilder>> originalPartitions;
+    private final List<Queue<PartitionBuilder>> originalPartitions;
```

Comment:
Fair. I'll take a stab at this.

Reply:
@jrwest `this.originalPartitions = new Queue<PartitionBuilder>[sources.length];` would actually be an illegal generic array creation, right?

Reply:
Thats what I get for not trying it first myself. Apologies for not thinking of that. Since I don't evidence showing this to be a problem, I'm fine leaving it as is.